src/api/wallet/WalletApiImpl.ts

```diff
@@ -0,0 +1,254 @@
+import { WalletApi, Network, WalletInfo, Command } from 'types'
+import BN from 'bn.js'
+import assert from 'assert'
+
+import WalletConnectProvider from '@walletconnect/web3-provider'
+
+import Web3 from 'web3'
+import { BlockHeader } from 'web3-eth'
+
+import {
+  getProvider,
+  getProviderState,
+  createSubscriptions,
+  Provider,
+  WalletConnectInits,
+  isWalletConnectProvider,
+  isMetamaskSubscriptions,
+  isWalletConnectSubscriptions,
+  isMetamaskProvider,
+  Subscriptions,
+} from '@gnosis.pm/dapp-ui'
+
+import { log, toBN } from 'utils'
+
+type OnChangeWalletInfo = (walletInfo: WalletInfo) => void
+
+// to track chain state, be that current account balance or token balance of that account
+// or any data on the chain belonging to that account
+// we need to refetch that data when
+// 1: network changes
+// 2: account changes
+// 3: new block is mined
+
+interface BlockchainUpdatePrompt {
+  account: string
+  chainId: number
+  blockHeader: BlockHeader | null
+}
+
+type BlockchainUpdatePromptCallback = (callback: (changedChainData: BlockchainUpdatePrompt) => void) => Command
+
+// provides subscription to blockhain updates for account/network/block
+const subscribeToBlockchainUpdate = ({
+  provider,
+  subscriptions,
+  web3,
+}: {
+  provider: Provider
+  subscriptions?: Subscriptions
+  web3: Web3
+}): BlockchainUpdatePromptCallback => {
+  const subs = subscriptions || createSubscriptions(provider)
+
+  let networkUpdate: (callback: (chainId: number) => void) => Command
+
+  if (isMetamaskSubscriptions(subs)) networkUpdate = (cb): Command => subs.onNetworkChanged(networkId => cb(+networkId))
+  if (isWalletConnectSubscriptions(subs)) networkUpdate = subs.onChainChanged
```

Comment:
I don't get why we have a lib for abstracting the wallet, but then we this wallet dependent logics here.

Reply:
The thing is what WalletConnect does and what MMask can do is significantly different.
Different enough that I opted to provide both options and leave it to the savvy user to choose one.

Subscriptions are very different and providers have to be handled differently. Same as I wrote above about `networkId`

If you want to only get a provider, then use `getProvider`, and deal with the result directly,
but if you want some helpers, then here you go.

Can do another abstraction , which would inevitably limit what can be done, but whatever. In say, a Mediator pattern, but later. If you wish.